src/experiment.js

```diff
@@ -0,0 +1,251 @@
+import Debug from 'debug'
+import Promise from 'bluebird'
+import assign from '101/assign'
+import find from '101/find'
+import hasProperties from '101/has-properties'
+import isString from '101/is-string'
+import isFunction from '101/is-function'
+import isObject from '101/is-object'
+import knuth_shuffle from 'knuth-shuffle'
+import { List, Map } from 'immutable'
+
+import MismatchError from './errors/mismatch-error'
+import Observation from './observation'
+import Result from './result'
+
+const debug = Debug('scientist:experiment')
+
+class Experiment {
+  constructor (name = 'experiment') {
+    debug('constructor')
+    this.name = name
+    this.enabled = true
+    this._behaviors = Map()
+    this._context = {}
+    this._ignores = List()
+    this._raise_on_mismatches = false
+  }
+
+  publish (result) {
+    debug('publish')
+    return Promise.resolve(true)
+  }
+
+  /**
+   * Define a Function to run before an experiment begins, if the experiment is
+   * enabled.
+   * @param {Function} fn Function to run.
+   */
+  before_run (fn) {
+    debug('before_run')
+    this._before_run_fn = fn
+  }
+
+  /**
+   * Define a function to clean a value for publishing or storing.
+   * @param {Function} fn Function to clean the value. Must take one argument,
+   *   the value to be cleaned.
+   */
+  clean (fn) {
+    debug('clean')
+    this._cleaner_fn = fn
+  }
+
+  /**
+   * Cleans a value with a configured clean block or just returns the value.
+   * @private
+   * @param {Object} value Value to be cleaned.
+   * @return {Object} Cleaned (if applicable) value.
+   */
+  clean_value (value) {
+    debug('clean_value')
```

Comment:
not _really_... the reason I have all those `debug` statements was to track the flow of things for a while. they can be expanded in the future, but not interested in doing that now